flink-connectors/flink-connector-datagen/src/main/java/org/apache/flink/connector/datagen/functions/FromElementsGeneratorFunction.java

```diff
@@ -0,0 +1,211 @@
+
+        try {
+            for (OUT element : elements) {
+                serializer.serialize(element, wrapper);
+            }
+        } catch (Exception e) {
+            throw new IOException("Serializing the source elements failed: " + e.getMessage(), e);
+        }
+        this.elementsSerialized = baos.toByteArray();
+    }
+
+    @Override
+    public void open(SourceReaderContext readerContext) throws Exception {
+        ByteArrayInputStream bais = new ByteArrayInputStream(elementsSerialized);
+        this.input = new DataInputViewStreamWrapper(bais);
+    }
+
+    @Override
+    public OUT map(Long nextIndex) throws Exception {
+        // Move iterator to the required position in case of failure recovery
+        while (numElementsEmitted < nextIndex) {
+            numElementsEmitted++;
+            tryDeserialize(serializer, input);
+        }
+        numElementsEmitted++;
+        return tryDeserialize(serializer, input);
+    }
+
+    private OUT tryDeserialize(TypeSerializer<OUT> serializer, DataInputView input)
+            throws IOException {
+        try {
+            return serializer.deserialize(input);
+        } catch (Exception e) {
+            throw new IOException(
+                    "Failed to deserialize an element from the source. "
+                            + "If you are using user-defined serialization (Value and Writable types), check the "
+                            + "serialization functions.\nSerializer is "
+                            + serializer,
+                    e);
+        }
+    }
+
+    @Override
+    public void setOutputType(TypeInformation<OUT> outTypeInfo, ExecutionConfig executionConfig) {
+        Preconditions.checkState(
+                elements != null,
+                "The output type should've been specified before shipping the graph to the cluster");
+        checkIterable(elements, outTypeInfo.getTypeClass());
+        TypeSerializer<OUT> newSerializer = outTypeInfo.createSerializer(executionConfig);
+        if (Objects.equals(serializer, newSerializer)) {
+            return;
+        }
+        serializer = newSerializer;
+        try {
+            serializeElements();
+        } catch (IOException ex) {
+            throw new UncheckedIOException(ex);
+        }
+    }
+
+    // ------------------------------------------------------------------------
+    //  Utilities
+    // ------------------------------------------------------------------------
+
+    /**
+     * Verifies that all elements in the collection are non-null, and are of the given class, or a
+     * subclass thereof.
+     *
+     * @param elements The collection to check.
+     * @param viewedAs The class to which the elements must be assignable to.
+     * @param <OUT> The generic type of the collection to be checked.
+     */
+    public static <OUT> void checkCollection(Collection<OUT> elements, Class<OUT> viewedAs) {
```

Comment:
Actually Im wondering why we need both this and checkiterable() 🤔

Reply:
>Actually Im wondering why we need both this and checkiterable() 🤔

Oh yeah, sure, my bad.

Reply:
>Why isnt this called in the constructor and instead called from the outside before calling the constructor?

This is just to keep it a bit in line with he old FromElementsFunction. Doing it in the construction would require passing TypeInformation that is otherwise not needed anywhere else in the class. Would you prefer to make the check internal?

Reply:
> Doing it in the construction would require passing TypeInformation that is otherwise not needed anywhere else in the class

Could we then skip the OutputTypeQueryable stuff? 🤔

Reply:
Yeah, that is what my first comment to the PR was about:

>The addition of the OutputTypeConfigurable was required to achieve feature parity and compatibility with the existing fromElements API ([issues.apache.org/jira/browse/FLINK-21386](https://issues.apache.org/jira/browse/FLINK-21386)). Alternatively, we could provide a fromElements method that accepts TypeInformation directly, rather than having to "fix" it with returns(), but that could potentially break user code.

I gave it a shot and I believe it is a better solution https://github.com/apache/flink/pull/23553/commits/2712c1813ca6420905e06b9e417de0eb61d586d9
I am not sure why [FLINK-21386](https://issues.apache.org/jira/browse/FLINK-21386) was addressed via returns(). 
My only concern is still that it will break user code for those who relied on "fixing" the serializer via that returns() call. We have both variants now - up to you.

Reply:
> My only concern is still that it will break user code for those who relied on "fixing" the serializer via that returns() call

That's really unfortunate; I guess we now have to keep it.

Reply:
Maybe it's actually even better to keep it because one should simply expect `returns` to work properly, even if the underlying implementation is pretty bad due to how the DataStream class works as a whole.

Reply:
✅ Added back the `returns()` support for backwards compatibility, but also kept the new method to pass the type directly (I think this one is more intuitive).